sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/storage/block/FileCompactBlockRepository.kt

```diff
@@ -0,0 +1,125 @@
+@file:Suppress("TooManyFunctions")
+
+package cash.z.ecc.android.sdk.internal.storage.block
+
+import androidx.annotation.VisibleForTesting
+import cash.z.ecc.android.sdk.internal.ext.createNewFileSuspend
+import cash.z.ecc.android.sdk.internal.ext.deleteRecursivelySuspend
+import cash.z.ecc.android.sdk.internal.ext.existsSuspend
+import cash.z.ecc.android.sdk.internal.ext.listFilesSuspend
+import cash.z.ecc.android.sdk.internal.ext.mkdirsSuspend
+import cash.z.ecc.android.sdk.internal.ext.readBytesSuspend
+import cash.z.ecc.android.sdk.internal.ext.renameToSuspend
+import cash.z.ecc.android.sdk.internal.ext.writeBytesSuspend
+import cash.z.ecc.android.sdk.internal.model.JniBlockMeta
+import cash.z.ecc.android.sdk.internal.repository.CompactBlockRepository
+import cash.z.ecc.android.sdk.internal.twig
+import cash.z.ecc.android.sdk.jni.RustBackend
+import cash.z.ecc.android.sdk.jni.RustBackendWelding
+import cash.z.ecc.android.sdk.model.BlockHeight
+import cash.z.ecc.android.sdk.model.ZcashNetwork
+import cash.z.wallet.sdk.rpc.CompactFormats.CompactBlock
+import java.io.File
+import java.util.Locale
+
+internal class FileCompactBlockRepository(
+    private val network: ZcashNetwork,
+    private val cacheDirectory: File,
+    private val rustBackend: RustBackendWelding
+) : CompactBlockRepository {
+
+    override suspend fun getLatestHeight() = rustBackend.getLatestHeight()
+
+    override suspend fun findCompactBlock(height: BlockHeight): CompactBlock? = // probably dont need
+        File(cacheDirectory, "blocks").listFilesSuspend()
+            ?.find { it.getBlockHeight() == height.value }
+            ?.let { CompactBlock.parseFrom(it.readBytesSuspend()) }
+
+    override suspend fun write(result: Sequence<CompactBlock>): Int {
+        var count = 0
+        val blocksDirectory = File(cacheDirectory, "blocks")
+        val metaDataBuffer = mutableListOf<JniBlockMeta>()
+
+        result.forEach { block ->
```

Comment:
I haven't Implemented this myself (yet) and therefore didn't measure it (but we are currently doing 100 block batches at most at the time), what would happen if this function receives an endless sequence of blocks? wouldn't it buffer them all and then try to write them at once through JNI?

Reply:
Pacu is right—we should have regular flushes of the buffer.  It could be as simple as a `forEachIndexed` and a modulo operator in the index.